EIPS/eip-5792.md

```diff
@@ -35,33 +35,41 @@ We also define one capability expression for use with the fourth method, which f
 
 ### `wallet_sendCalls`
 
-Requests that a wallet submits a batch of calls. `from` and `chainId` are top-level properties rather than per-call properties because all calls should be sent from the same sender and on the same chain, identified by [EIP-155](./eip-155.md) integers expressed in hexadecimal notation. The items in the `calls` field are only those that are shared by all transaction types. Any other fields that a wallet may need to submit a transaction should be handled by the wallet.
+Requests that a wallet submits a batch of calls. `from` and `chainId` are top-level properties rather than per-call properties because all calls should be sent from the same sender and on the same chain, identified by [EIP-155](./eip-155.md) integers expressed in hexadecimal notation.
+The items in the `calls` field are simple `{to, data, value}` tuples.
 
 The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.
 
+Each capability defined in the "capabilities" member can define global or call specific fields.
+These fields are set inside this capability's entry in the `capabilitiesData` object.
+Each entity in the `calls` field may contain an optional `capabilitiesData` object.
+This object allows the applications to attach a capability-specific metadata to individual calls.
+
 The wallet:
 
 * MUST send the calls in the order specified in the request
-* MUST send the calls on the same chain identified by the call's `chainId` property
+* MUST send the calls on the same chain identified by the request's `chainId` property
 * MUST NOT await for any calls to be finalized to complete the batch
 * MUST NOT send any calls from the request if the user rejects the request
 * MAY revert all calls if any call fails
 * MAY send all calls as part of one or more transactions, depending on wallet capability
 * SHOULD stop executing the calls if any call fails
 * MAY reject the request if the from address does not match the enabled account
 * MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially
+* MUST reject the request if it contains a `capability` that is not supported by the wallet
```

Comment:
Hmm I wonder if this is best 🤔 . E.g. why not allow specifying a paymaster which the app would like to be used but is not a necessity.

Reply:
How a wallet can tell if the cap (e.g. paymaster) is required by this app or optional?
The app can always query the wallet caps first, to decide how to continue.

Reply:
capability specs should specify whether they're required or not / how optional ones are handled. paymaster, for example, notes that the app must not assume the provided paymaster service is the one that ends up being used.

so if an app wants the capability to be required, the capability spec should include a `required` field, and that is how a wallet can tell if the capability is required by the app. i dont think we should blanket all capabilities to be required by default.

Reply:
There are two issues raised here:

1. App providing a "capability" that is *not supported* by the wallet:

There is a `wallet_getCapabilities` method that provides the application with a list of *supported* "capabilities".
Shouldn't the app that ignores that list and submits a `wallet_sendCalls` with a "capability" that is not supported be considered to behave incorrectly? I think it should reject the call, which would prevent dapps from yoloing their `sendCalls` requests.

2. Wallet discretion to use or ignore a *supported* "capability" requested by the dapp:

As the API defined here deals with real value and assets, making capabilities `required` by default makes a lot of sense to me. If using a paymaster for a transaction is indeed optional, it is trivial to specify `optional: true`, but forgetting to set `required: true` for a security-oriented capability may be catastrophic. 

What do you think?

Reply:
So @forshtat is the conclusion here that each capability *could* have an `optional: ` flag? But passing a capability the wallet doesn't have would be hard failure?

Reply:
> So @forshtat is the conclusion here that each capability _could_ have an `optional: ` flag? But passing a capability the wallet doesn't have would be hard failure?

In my personal opinion, yes, this seems pretty reasonable.